core/invoker/src/main/scala/whisk/core/container/HttpUtils.scala

```diff
@@ -0,0 +1,135 @@
+import org.apache.http.client.methods.HttpPost
+import org.apache.http.client.methods.HttpRequestBase
+import org.apache.http.client.utils.URIBuilder
+import org.apache.http.conn.HttpHostConnectException
+import org.apache.http.entity.StringEntity
+import org.apache.http.impl.client.HttpClientBuilder
+import org.apache.http.util.EntityUtils
+
+import spray.json.JsObject
+import whisk.core.entity.ActivationResponse._
+import whisk.core.entity.ByteSize
+import whisk.core.entity.size.SizeLong
+
+/**
+ * This HTTP client is used only in the invoker to communicate with the action container.
+ * It allows to POST a JSON object and receive JSON object back; that is the
+ * content type and the accept headers are both 'application/json.
+ * The reason we still use this class for the action container is a mysterious hang
+ * in the Akka http client where a future fails to properly timeout and we have not
+ * determined why that is.
+ *
+ * @param hostname the host name
+ * @param timeoutMsec the timeout in msecs to wait for a response
+ */
+protected[container] class HttpUtils(
+    hostname: String,
+    timeout: FiniteDuration,
+    maxResponse: ByteSize) {
+
+    def close = Try(connection.close)
+
+    /**
+     * Posts to hostname/endpoint the given JSON object.
+     * Waits up to timeout before aborting on a good connection.
+     * If the endpoint is not ready, retry up to timeout.
+     * Every retry reduces the available timeout so that this method should not
+     * wait longer than the total timeout (within a small slack allowance).
+     *
+     * @param endpoint the path the api call relative to hostname
+     * @param body the json object to post
+     * @return Left(Error Message) or Right(Status Code, Response as UTF-8 String)
+     */
+    def post(endpoint: String, body: JsObject): Either[ContainerConnectionError, ContainerResponse] = {
+        val entity = new StringEntity(body.compactPrint, StandardCharsets.UTF_8)
+        entity.setContentType("application/json")
+
+        val request = new HttpPost(baseUri.setPath(endpoint).build)
+        request.addHeader(HttpHeaders.ACCEPT, "application/json")
+        request.setEntity(entity)
+
+        execute(request, timeout.toMillis.toInt)
+    }
+
+    private def execute(request: HttpRequestBase, timeoutMsec: Integer): Either[ContainerConnectionError, ContainerResponse] = {
+        var response: CloseableHttpResponse = null
+        try {
+            response = connection.execute(request)
+            Try(response.getEntity).filter(_ != null).map { entity =>
+                val statusCode = response.getStatusLine.getStatusCode
+                entity.getContentLength match {
+                    case length if length != -1 =>
+                        if (length <= maxResponseBytes) {
+                            val str = EntityUtils.toString(entity, StandardCharsets.UTF_8)
+                            Right(ContainerResponse(statusCode, str))
```

Comment:
As this is tightly coupled to spray anyway, we might as well use `StatusCode` instead of an `Int` here for some extra typedness and easier interchangeability later?